main/autogen/autoloader.cc

```diff
@@ -388,31 +390,26 @@ void DefTreeBuilder::collapseSameFileDefs(const core::GlobalState &gs, const Aut
     }
 }
 
-void AutoloadWriter::writeAutoloads(const core::GlobalState &gs, const AutoloaderConfig &alCfg, const std::string &path,
-                                    const DefTree &root) {
-    UnorderedSet<string> toDelete; // Remove from this set as we write files
-    if (FileOps::exists(path)) {
-        vector<string> existingFiles = FileOps::listFilesInDir(path, {".rb"}, true, {}, {});
-        toDelete.insert(make_move_iterator(existingFiles.begin()), make_move_iterator(existingFiles.end()));
-    }
-    write(gs, alCfg, path, toDelete, root);
-    for (const auto &file : toDelete) {
-        FileOps::removeFile(file);
-    }
-}
-
-void AutoloadWriter::write(const core::GlobalState &gs, const AutoloaderConfig &alCfg, const std::string &path,
-                           UnorderedSet<std::string> &toDelete, const DefTree &node) {
+namespace {
+struct RenderAutoloadTask {
+    string filePath;
+    const DefTree &node;
+};
+
+// This function has two duties:
+// * It creates autoload rendering tasks which will occur in a later parallel phase.
+// * It creates subdirectories when needed, as they are required to write the autoloader output.
+void populateAutoloadTasksAndCreateDirectories(const core::GlobalState &gs, vector<RenderAutoloadTask> &tasks,
+                                               const AutoloaderConfig &alCfg, string_view path, const DefTree &node) {
     string name = node.root() ? "root" : node.name().show(gs);
     string filePath = join(path, fmt::format("{}.rb", name));
     if (!alCfg.packagedAutoloader || !node.root()) {
-        FileOps::writeIfDifferent(filePath, node.renderAutoloadSrc(gs, alCfg));
```

Comment:
Rendering the autoload src and writing to disk dominated this formally serial part of autogen. This PR defers that work to worker threads.